// aws-cdk/lib/cli/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trace = exports.debug = exports.info = exports.warn = exports.result = exports.error = void 0;
exports.formatMessage = formatMessage;
/**
 * Internal helper that processes log inputs into a consistent format.
 * Handles string interpolation, format strings, and object parameter styles.
 * Applies optional styling and prepares the final message for logging.
 */
function formatMessage(msg, category = 'TOOLKIT') {
    var _a;
    return {
        time: new Date(),
        level: msg.level,
        action: msg.action,
        code: (_a = msg.code) !== null && _a !== void 0 ? _a : defaultMessageCode(msg.level, category),
        message: msg.message,
        data: msg.data,
    };
}
/**
 * Build a message code from level and category. The code must be valid for this function to pass.
 * Otherwise it returns a ToolkitError.
 */
function defaultMessageCode(level, category = 'TOOLKIT') {
    const levelIndicator = level === 'error' ? 'E' :
        level === 'warn' ? 'W' :
            'I';
    return `CDK_${category}_${levelIndicator}0000`;
}
/**
 * Creates an error level message.
 * Errors must always have a unique code.
 */
const error = (action, message, code, payload) => {
    return formatMessage({
        level: 'error',
        action,
        code,
        message,
        data: payload,
    });
};
exports.error = error;
/**
 * Creates a result level message and represents the most important message for a given action.
 *
 * They should be used sparsely, with an action usually having no or exactly one result.
 * However actions that operate on Cloud Assemblies might include a result per Stack.
 * Unlike other messages, results must always have a code and a payload.
 */
const result = (action, message, code, payload) => {
    return formatMessage({
        level: 'result',
        action,
        code,
        message,
        data: payload,
    });
};
exports.result = result;
/**
 * Creates a warning level message.
 */
const warn = (action, message, code, payload) => {
    return formatMessage({
        level: 'warn',
        action,
        code,
        message,
        data: payload,
    });
};
exports.warn = warn;
/**
 * Creates an info level message.
 */
const info = (action, message, code, payload) => {
    return formatMessage({
        level: 'info',
        action,
        code,
        message,
        data: payload,
    });
};
exports.info = info;
/**
 * Creates a debug level message.
 */
const debug = (action, message, code, payload) => {
    return formatMessage({
        level: 'debug',
        action,
        code,
        message,
        data: payload,
    });
};
exports.debug = debug;
/**
 * Creates a trace level message.
 */
const trace = (action, message, code, payload) => {
    return formatMessage({
        level: 'trace',
        action,
        code,
        message,
        data: payload,
    });
};
exports.trace = trace;
