// aws-cdk/lib/legacy-exports-source.js
"use strict";
// This is a barrel export file, of all known symbols that are imported by users from the `aws-cdk` package.
// Importing these symbols was never officially supported, but here we are.
// In order to preserver backwards-compatibly for these users, we re-export and preserve them as explicit subpath exports.
// See https://github.com/aws/aws-cdk/pull/33021 for more information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsCliCompatible = exports.CredentialPlugins = exports.CfnEvaluationException = exports.cached = exports.availableInitTemplates = exports.versionNumber = exports.latestVersionIfHigher = exports.rootDir = exports.Deployments = exports.deepMerge = exports.lowerCaseFirstCharacter = exports.describe = exports.command = exports.aliases = exports.enableTracing = exports.formatAsBanner = exports.leftPad = exports.RequireApproval = exports.execProgram = exports.CloudExecutable = exports.Bootstrapper = exports.Settings = exports.PROJECT_CONTEXT = exports.Configuration = exports.Command = exports.contentHash = exports.PluginHost = exports.SdkProvider = exports.exec = exports.cli = exports.deployStack = exports.partition = exports.numberFromBool = exports.isEmpty = exports.isArray = exports.ifDefined = exports.flatten = exports.deepClone = void 0;
// Note: All type exports are in `legacy-exports.ts`
__exportStar(require("./legacy-logging-source"), exports);
var util_1 = require("./util");
Object.defineProperty(exports, "deepClone", { enumerable: true, get: function () { return util_1.deepClone; } });
Object.defineProperty(exports, "flatten", { enumerable: true, get: function () { return util_1.flatten; } });
Object.defineProperty(exports, "ifDefined", { enumerable: true, get: function () { return util_1.ifDefined; } });
Object.defineProperty(exports, "isArray", { enumerable: true, get: function () { return util_1.isArray; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return util_1.isEmpty; } });
Object.defineProperty(exports, "numberFromBool", { enumerable: true, get: function () { return util_1.numberFromBool; } });
Object.defineProperty(exports, "partition", { enumerable: true, get: function () { return util_1.partition; } });
var deploy_stack_1 = require("./api/deployments/deploy-stack");
Object.defineProperty(exports, "deployStack", { enumerable: true, get: function () { return deploy_stack_1.deployStack; } });
var cli_1 = require("./cli/cli");
Object.defineProperty(exports, "cli", { enumerable: true, get: function () { return cli_1.cli; } });
Object.defineProperty(exports, "exec", { enumerable: true, get: function () { return cli_1.exec; } });
var aws_auth_1 = require("./api/aws-auth");
Object.defineProperty(exports, "SdkProvider", { enumerable: true, get: function () { return aws_auth_1.SdkProvider; } });
var plugin_1 = require("./api/plugin");
Object.defineProperty(exports, "PluginHost", { enumerable: true, get: function () { return plugin_1.PluginHost; } });
var content_hash_1 = require("./util/content-hash");
Object.defineProperty(exports, "contentHash", { enumerable: true, get: function () { return content_hash_1.contentHash; } });
var user_configuration_1 = require("./cli/user-configuration");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return user_configuration_1.Command; } });
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return user_configuration_1.Configuration; } });
Object.defineProperty(exports, "PROJECT_CONTEXT", { enumerable: true, get: function () { return user_configuration_1.PROJECT_CONTEXT; } });
var settings_1 = require("./api/settings");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
var bootstrap_1 = require("./api/bootstrap");
Object.defineProperty(exports, "Bootstrapper", { enumerable: true, get: function () { return bootstrap_1.Bootstrapper; } });
var cloud_executable_1 = require("./api/cxapp/cloud-executable");
Object.defineProperty(exports, "CloudExecutable", { enumerable: true, get: function () { return cloud_executable_1.CloudExecutable; } });
var exec_1 = require("./api/cxapp/exec");
Object.defineProperty(exports, "execProgram", { enumerable: true, get: function () { return exec_1.execProgram; } });
var diff_1 = require("./diff");
Object.defineProperty(exports, "RequireApproval", { enumerable: true, get: function () { return diff_1.RequireApproval; } });
var string_manipulation_1 = require("./api/util/string-manipulation");
Object.defineProperty(exports, "leftPad", { enumerable: true, get: function () { return string_manipulation_1.leftPad; } });
var console_formatters_1 = require("./cli/util/console-formatters");
Object.defineProperty(exports, "formatAsBanner", { enumerable: true, get: function () { return console_formatters_1.formatAsBanner; } });
var tracing_1 = require("./api/aws-auth/tracing");
Object.defineProperty(exports, "enableTracing", { enumerable: true, get: function () { return tracing_1.setSdkTracing; } });
var docs_1 = require("./commands/docs");
Object.defineProperty(exports, "aliases", { enumerable: true, get: function () { return docs_1.aliases; } });
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return docs_1.command; } });
Object.defineProperty(exports, "describe", { enumerable: true, get: function () { return docs_1.describe; } });
var common_1 = require("./api/hotswap/common");
Object.defineProperty(exports, "lowerCaseFirstCharacter", { enumerable: true, get: function () { return common_1.lowerCaseFirstCharacter; } });
var objects_1 = require("./util/objects");
Object.defineProperty(exports, "deepMerge", { enumerable: true, get: function () { return objects_1.deepMerge; } });
var deployments_1 = require("./api/deployments");
Object.defineProperty(exports, "Deployments", { enumerable: true, get: function () { return deployments_1.Deployments; } });
var directories_1 = require("./util/directories");
Object.defineProperty(exports, "rootDir", { enumerable: true, get: function () { return directories_1.rootDir; } });
var version_1 = require("./cli/version");
Object.defineProperty(exports, "latestVersionIfHigher", { enumerable: true, get: function () { return version_1.latestVersionIfHigher; } });
Object.defineProperty(exports, "versionNumber", { enumerable: true, get: function () { return version_1.versionNumber; } });
var init_1 = require("./init");
Object.defineProperty(exports, "availableInitTemplates", { enumerable: true, get: function () { return init_1.availableInitTemplates; } });
var cached_1 = require("./api/aws-auth/cached");
Object.defineProperty(exports, "cached", { enumerable: true, get: function () { return cached_1.cached; } });
var evaluate_cloudformation_template_1 = require("./api/evaluate-cloudformation-template");
Object.defineProperty(exports, "CfnEvaluationException", { enumerable: true, get: function () { return evaluate_cloudformation_template_1.CfnEvaluationException; } });
var credential_plugins_1 = require("./api/aws-auth/credential-plugins");
Object.defineProperty(exports, "CredentialPlugins", { enumerable: true, get: function () { return credential_plugins_1.CredentialPlugins; } });
var awscli_compatible_1 = require("./api/aws-auth/awscli-compatible");
Object.defineProperty(exports, "AwsCliCompatible", { enumerable: true, get: function () { return awscli_compatible_1.AwsCliCompatible; } });
