// aws-cdk/lib/toolkit/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextProviderError = exports.AssemblyError = exports.AuthenticationError = exports.ToolkitError = void 0;
const TOOLKIT_ERROR_SYMBOL = Symbol.for('@aws-cdk/toolkit.ToolkitError');
const AUTHENTICATION_ERROR_SYMBOL = Symbol.for('@aws-cdk/toolkit.AuthenticationError');
const ASSEMBLY_ERROR_SYMBOL = Symbol.for('@aws-cdk/toolkit.AssemblyError');
const CONTEXT_PROVIDER_ERROR_SYMBOL = Symbol.for('@aws-cdk/toolkit.ContextProviderError');
/**
 * Represents a general toolkit error in the AWS CDK Toolkit.
 */
class ToolkitError extends Error {
    /**
     * Determines if a given error is an instance of ToolkitError.
     */
    static isToolkitError(x) {
        return x !== null && typeof (x) === 'object' && TOOLKIT_ERROR_SYMBOL in x;
    }
    /**
     * Determines if a given error is an instance of AuthenticationError.
     */
    static isAuthenticationError(x) {
        return this.isToolkitError(x) && AUTHENTICATION_ERROR_SYMBOL in x;
    }
    /**
     * Determines if a given error is an instance of AssemblyError.
     */
    static isAssemblyError(x) {
        return this.isToolkitError(x) && ASSEMBLY_ERROR_SYMBOL in x;
    }
    /**
     * Determines if a given error is an instance of AssemblyError.
     */
    static isContextProviderError(x) {
        return this.isToolkitError(x) && CONTEXT_PROVIDER_ERROR_SYMBOL in x;
    }
    constructor(message, type = 'toolkit') {
        super(message);
        Object.setPrototypeOf(this, ToolkitError.prototype);
        Object.defineProperty(this, TOOLKIT_ERROR_SYMBOL, { value: true });
        this.name = new.target.name;
        this.type = type;
    }
}
exports.ToolkitError = ToolkitError;
/**
 * Represents an authentication-specific error in the AWS CDK Toolkit.
 */
class AuthenticationError extends ToolkitError {
    constructor(message) {
        super(message, 'authentication');
        Object.setPrototypeOf(this, AuthenticationError.prototype);
        Object.defineProperty(this, AUTHENTICATION_ERROR_SYMBOL, { value: true });
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * Represents an authentication-specific error in the AWS CDK Toolkit.
 */
class AssemblyError extends ToolkitError {
    constructor(message) {
        super(message, 'assembly');
        Object.setPrototypeOf(this, AssemblyError.prototype);
        Object.defineProperty(this, ASSEMBLY_ERROR_SYMBOL, { value: true });
    }
}
exports.AssemblyError = AssemblyError;
/**
 * Represents an error originating from a Context Provider
 */
class ContextProviderError extends ToolkitError {
    constructor(message) {
        super(message, 'context-provider');
        Object.setPrototypeOf(this, ContextProviderError.prototype);
        Object.defineProperty(this, CONTEXT_PROVIDER_ERROR_SYMBOL, { value: true });
    }
}
exports.ContextProviderError = ContextProviderError;
